### ISSUE TITLE
Memory leak with sensitivities

### ISSUE THREAD

**Issue by DavAug** [myokit/myokit#870]
Very similar to #867 , there seem to be residual memory allocations from the sensitivities.

Toy problem:
```
import resource
import myokit

m, p, _ = myokit.load('example')

# Enable sensitivities
states = [var.qname() for var in m.states()]
params = []
for var in m.variables(const=True):
    # Sometimes constants are derived from parameters
    if not var.is_literal():
        continue
    params.append(var.qname())
s = myokit.Simulation(m, p, (states, params))

s.run(10, log=myokit.LOG_NONE)

u1 = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
for i in range(100):
    s.run(1, log=myokit.LOG_NONE)
u2 = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

print(u2 - u1)
```

Without sensitivities this returns ``0`` (as we would expect). With sensitivities this returns 

```
(venv) ➜  myokit git:(main) ✗ python todelete.py
WARNING:myokit:Using development version of Myokit (1.33.4.dev).
311296
```

**Comment by MichaelClerx** [comment]
> @MichaelClerx Is it possible that
> 
> ```
> sy = N_VCloneVectorArray(model->ns_independents, y);
> ```
> 
> and
> 
> ```
> if (sy != NULL) { N_VDestroyVectorArray(sy, model->ns_independents); sy = NULL; }
> ```
> 
> are not counterparts of each other?

It's possible, but you can check that in the sundials documentation :D
I think it's correct though, and we would get an error otherwise

**Comment by MichaelClerx** [comment]
I've added a slightly more complicated test script here: https://github.com/MichaelClerx/myokit/tree/main/myokit/tests/memory
And just added your testcase (sort of). With the default settings it shows no change. If I make it more like your toy problem (protocol from example file, no s.reset(), same run durations) I get a single memory increase. For definite memory leaks I usually saw 2 or 3 (try it by commenting out a free() line).

I have no idea why this seems to be non-deterministic 🤷🏻‍♂️ 

Maybe a good idea to also ask Martin or Fergus what they think is the best way to find these memory leaks? (And check that they actually are memory leaks, not e.g. some Python list somewhere increasing in size).

**Comment by DavAug** [comment]
> > @MichaelClerx Is it possible that
> > ```
> > sy = N_VCloneVectorArray(model->ns_independents, y);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and
> > ```
> > if (sy != NULL) { N_VDestroyVectorArray(sy, model->ns_independents); sy = NULL; }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > are not counterparts of each other?
> 
> It's possible, but you can check that in the sundials documentation :D I think it's correct though, and we would get an error otherwise

Yes it seem correct. I just looked it up. I am just wondering what could make the memory usage grow if not that :D

**Comment by MichaelClerx** [comment]
> I am just wondering what could make the memory usage grow if not that :D

This is a pretty good question. I'm just looking at the code to see if I can spot anything.

- https://github.com/MichaelClerx/myokit/blob/main/myokit/_sim/cvodessim.py#L332-L339 Could be better to change values of existing list entries instead of creating new lists and having the old ones garbage collected
- https://github.com/MichaelClerx/myokit/blob/main/myokit/_sim/cvodessim.py#L529-L530 Could skip this step if log=LOG_NONE? (And then return `None` instead of an empty list?)
- https://github.com/MichaelClerx/myokit/blob/main/myokit/_sim/cvodessim.py#L573 Could skip this step if log=LOG_NONE? (And then return `None` instead of an empty list?)

Some more creation going on here that can't be avoided: https://github.com/MichaelClerx/myokit/blob/main/myokit/_sim/cvodessim.py#L584-L591

In general though, these cases should all eventually be garbage collected. Python grabs a big chunk of memory, and then makes lists, floats, whatever in that chunk. GC doesn't free that chunk, it just makes it available for re-use. So we might often see Python objects creation not affect the total memory use, and then sometimes for whatever reason it will.

I have no idea why we can't detect every single change in memory that the compiled C code uses.

Something you could do to check if it's just python memory is invoke `gc.collect()` during every iteration of your loop. This could make things better by reducing the chance that Python increases the memory it has reserved. It could also make things worse, if the GC process itself is so complicated that it needs more memory :D

**Comment by MichaelClerx** [comment]
> But I have no idea why we can't detect every single change in memory that the compiled C code uses.

https://stackoverflow.com/questions/6005333/problem-with-free-on-structs-in-c-it-doesnt-reduce-memory-usage

https://stackoverflow.com/questions/5994543/problem-usage-memory-in-c/5994981#5994981

Okay so to a degree C does the same thing that Python does and keeps the memory freed by `free()` so that it can give it back to you when you next call `malloc()`.

Something I've seen people do is run a few iterations, say 10, to get all the various caches up to good sizes before starting any profiling

**Comment by DavAug** [comment]
For example the ``udata`` and the ``pbar`` parameters are created when the sensitivities are enabled.
```
udata = (UserData)malloc(sizeof *udata);
```
and 
```
udata->p = (realtype*)malloc(sizeof(realtype) * model->ns_independents);
```

But I guess because we free them in sim_clean there is no way that those could give rise to this behaviour?
```
/* User data and parameter scale array*/
        free(pbar);
        if (udata != NULL) {
            free(udata->p);
            free(udata); udata = NULL;
        }
```

**Comment by DavAug** [comment]
Hmm ok, that sounds convincing to me. I think we might be able to use pympler to count the Python objects, so then we should see whether we the sensitivities keep creating more lists. I have done this before, but I can’t remember whether I only checked the simulation without sensitivities and that was why I didn’t see any more Python objects being created.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by MichaelClerx** [comment]
I reworked your pympler thing into this: https://github.com/MichaelClerx/myokit/blob/main/myokit/tests/memory/decref.py

But I don't think pympler would show it either. What I think happens is that Python has some chunk of memory, say 1 MB, of which 20% is in use, then we run a simulation and 40% is in use, then we run another and 60% is in use, then it does GC but with some fragmentation etc. so still 21% is in use, but we can keep running and never see an increase (it stays at 1MB). Now if the simulation uses more memory (like the sensitivity one), then there's more chance of the 1MB being exceeded during the run, and before the next GC. So maybe Python will raise its memory from 1MB to 1.2MB. Then eventually GC and go back to using 20% of 1MB (but mem use still shows as 1.2MB). So Pympler would correctly say that there is no leak, but your process manager would see an increase in used memory at some point

**Comment by MichaelClerx** [comment]
As an aside, a lot of places I'm reading say using subprocesses is a good way to avoid this, as once a subprocess halts its memory is released back to the OS. But I assume you're already using multiprocessing for your big hierarchical thing?

**Comment by DavAug** [comment]
> I reworked your pympler thing into this: https://github.com/MichaelClerx/myokit/blob/main/myokit/tests/memory/decref.py
> 
> But I don't think pympler would show it either. What I think happens is that Python has some chunk of memory, say 1 MB, of which 20% is in use, then we run a simulation and 40% is in use, then we run another and 60% is in use, then it does GC but with some fragmentation etc. so still 21% is in use, but we can keep running and never see an increase (it stays at 1MB). Now if the simulation uses more memory (like the sensitivity one), then there's more chance of the 1MB being exceeded during the run, and before the next GC. So maybe Python will raise its memory from 1MB to 1.2MB. Then eventually GC and go back to using 20% of 1MB (but mem use still shows as 1.2MB). So Pympler would correctly say that there is no leak, but your process manager would see an increase in used memory at some point

Hmm ok, I am still trying to wrap my had around this. What I don't understand is that usually we don't get this memory fragmentation from python do we? So this behaviour must be connected to our c script. If we assume that we correctly handle the dynamically created memory allocations, the error must come from the PyObjects in the c code and similar to what you are saying those PyObjects are somehow not properly detected by pythons garbage collector. Do you know whether there is a way to "free" PyObjects explicitly in the c script? Then we could just call this method at the end of every simulation to avoid any fragmentation?

(Btw in support of your list creation theory, I went back to the non-sensitivity simulation, and for 20 runs versus 10000 runs we get again differences in the memory usage)

```
import resource
import myokit

m, p, _ = myokit.load('example')

# # Enable sensitivities
# states = [var.qname() for var in m.states()]
# params = []
# for var in m.variables(const=True):
#     # Sometimes constants are derived from parameters
#     if not var.is_literal():
#         continue
#     params.append(var.qname())
# s = myokit.Simulation(m, p, (states, params))
s = myokit.Simulation(m, p)

for i in range(20):
    s.reset()
    s.run(10, log=myokit.LOG_NONE)

u1 = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
for i in range(10000):
    s.reset()
    s.run(10, log=myokit.LOG_NONE)
u2 = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

print(u2 - u1)

``` 

```
import resource
import myokit

m, p, _ = myokit.load('example')
s = myokit.Simulation(m, p)

for i in range(20):
    s.reset()
    s.run(10, log=myokit.LOG_NONE)

u1 = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
for i in range(10000):
    s.reset()
    s.run(10, log=myokit.LOG_NONE)
u2 = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

print(u2 - u1)
```

**Comment by DavAug** [comment]
> As an aside, a lot of places I'm reading say using subprocesses is a good way to avoid this, as once a subprocess halts its memory is released back to the OS. But I assume you're already using multiprocessing for your big hierarchical thing?

Interesting idea! I am currently not parallelising the simulation runs, because in my initial attempt I didn't gain any speed up (but I also initialised a new streamlined evaluation of the myokit model at each posterior evaluation, which is probably why it was so slow).

**Comment by MichaelClerx** [comment]
> What I don't understand is that usually we don't get this memory fragmentation from python do we?
> ... So this behaviour must be connected to our c script. 

You do! For example, say you created an int, then a list, then an int, then a list, etc. until you had 10 ints and 10 lists.
Now you `del(a, b, c, ...)` delete all the ints. Technically Python has 10 int spaces free. But if you ask for a new list capable of holding 10 ints, it can't use those bits of memory, so it will have to use memory after the 10th list to store list 11.

(I think this might be further complicated by the fact that garbage collection only happens occasionally, so if you create a list, delete the list, and then create another list, you might still use list-size * 2 in memory, because Python hasn't done the work to realise that the space occupied by list 1 is available for re-use.)

Your pympler is very good, because it checks all the stuff that is known to exist, so it will tell you if **you** are leaking objects. It won't tell you if Python has increased its total memory footprint. Top, or any process manager, or the `resource` module will tell you the total memory footprint, but it won't tell you how much of that is being used.

It's also not Python's goal to be optimally memory efficient, so we shouldn't expect it to use exactly the amount of memory we've worked out that we need. It's goal is to find a good trade-off between low memory use and speed (memory management takes time). The same holds for C's free() and malloc().

> If we assume that we correctly handle the dynamically created memory allocations, the error must come from the PyObjects in the c code and similar to what you are saying those PyObjects are somehow not properly detected by pythons garbage collector. Do you know whether there is a way to "free" PyObjects explicitly in the c script? Then we could just call this method at the end of every simulation to avoid any fragmentation?

The pympler tool would show us if the C code was leaking PyObjects (that's how you found the previous leak!). The problem there was exactly that I'd created a new PyObject but not "freed" it. (Creating it increased its reference count to 1, and then I forgot to reduce the reference count by 1 when I was finished with it, so Python didn't know that it was allowed tidy it up).

**Comment by DavAug** [comment]
> > What I don't understand is that usually we don't get this memory fragmentation from python do we?
> > ... So this behaviour must be connected to our c script.
> 
> You do! For example, say you created an int, then a list, then an int, then a list, etc. until you had 10 ints and 10 lists. Now you `del(a, b, c, ...)` delete all the ints. Technically Python has 10 int spaces free. But if you ask for a new list capable of holding 10 ints, it can't use those bits of memory, so it will have to use memory after the 10th list to store list 11.
> 
> (I think this might be further complicated by the fact that garbage collection only happens occasionally, so if you create a list, delete the list, and then create another list, you might still use list-size * 2 in memory, because Python hasn't done the work to realise that the space occupied by list 1 is available for re-use.)
> 
> Your pympler is very good, because it checks all the stuff that is known to exist, so it will tell you if **you** are leaking objects. It won't tell you if Python has increased its total memory footprint. Top, or any process manager, or the `resource` module will tell you the total memory footprint, but it won't tell you how much of that is being used.
> 
> It's also not Python's goal to be optimally memory efficient, so we shouldn't expect it to use exactly the amount of memory we've worked out that we need. It's goal is to find a good trade-off between low memory use and speed (memory management takes time). The same holds for C's free() and malloc().
> 
> > If we assume that we correctly handle the dynamically created memory allocations, the error must come from the PyObjects in the c code and similar to what you are saying those PyObjects are somehow not properly detected by pythons garbage collector. Do you know whether there is a way to "free" PyObjects explicitly in the c script? Then we could just call this method at the end of every simulation to avoid any fragmentation?
> 
> The pympler tool would show us if the C code was leaking PyObjects (that's how you found the previous leak!). The problem there was exactly that I'd created a new PyObject but not "freed" it. (Creating it increased its reference count to 1, and then I forgot to reduce the reference count by 1 when I was finished with it, so Python didn't know that it was allowed tidy it up).

Ahh ok thanks for explaining that to me. (Hopefully) my last question :D: it sounds like Python might not be optimally controlling the memory usage, but what I don't understand is how evaluating the simulation run for 12 hours in a row can lead to 100% memory occupation? I mean, I can understand how the garbage collection design prevents you from being optimal and you can end up having occasionally fragmented memory allocations, but since we produce and discard objects every half second or so, I would think that the increase in memory shouldn't continue for 12 hours? Especially since the memory usage appears to increase linearly (I would understand some kind of logistic increase that levels off to some value).
In other words, shouldn't the garbage collector be able to pick up on unused memory allocations eventually, even if it might take a couple more rounds than necessary?

**Comment by DavAug** [comment]
Ok, this seems to suggest that the memory usage does indeed level off over time. That doesn't really help me with my problem, but at least we may be able to close this issue and there really is no further leak from myokit.

Increase in memory usage for list creation:
```
import resource

import matplotlib.pyplot as plt


def run():
    return [[x] for x in range(1000)]


runs = 100
usage = []
for run_id in range(runs):
    for i in range(1000):
        run()
    u = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    usage.append(u)

plt.plot(list(range(runs)), usage)
plt.show()
```

Increase in memory usage for simulation without sensitivities:
```
import resource
import myokit

import matplotlib.pyplot as plt


m, p, _ = myokit.load('example')
s = myokit.Simulation(m, p)


runs = 100
usage = []
for run_id in range(runs):
    for i in range(1000):
        s.reset()
        s.run(10, log=myokit.LOG_NONE)
    u = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    usage.append(u)

plt.plot(list(range(runs)), usage)
plt.show()
```

Incease of memory usage with sensitivities:
```
import resource
import myokit

import matplotlib.pyplot as plt


m, p, _ = myokit.load('example')
# Enable sensitivities
states = [var.qname() for var in m.states()]
params = []
for var in m.variables(const=True):
    # Sometimes constants are derived from parameters
    if not var.is_literal():
        continue
    params.append(var.qname())
s = myokit.Simulation(m, p, (states, params))


runs = 100
usage = []
for run_id in range(runs):
    for i in range(1000):
        s.reset()
        s.run(10, log=myokit.LOG_NONE)
    u = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    usage.append(u)

plt.plot(list(range(runs)), usage)
plt.show()

```

**Comment by MichaelClerx** [comment]
> but what I don't understand is how evaluating the simulation run for 12 hours in a row can lead to 100% memory occupation? I

OK that sounds exactly like a memory leak 😅😅😅

**Comment by MichaelClerx** [comment]
OK, so...

1. There's a memory leak in Myokit and we're testing correctly but misinterpreting the result,
2. There's a memory leak in Myokit but our tests aren't picking it up
3. There's a memory leak somewhere else

I feel like we've done a reasonable job of discounting 1, so I'll try and look into 2 a bit :D

**Comment by DavAug** [comment]
> > but what I don't understand is how evaluating the simulation run for 12 hours in a row can lead to 100% memory occupation?
> 
> OK that sounds exactly like a memory leak 😅😅😅
Haha I know! :D that is why I was looking for one all the time.

But interestingly, I am currently simulating a model using multiprocessing and the memory issues are indeed avoided. I can see how the memory is cleared after each subprocesses has finished its execution. So in other words, I can see how the memory increases during evaluation, but once the job is finished the process memory is killed and I don't run out of memory anymore.

**Comment by MichaelClerx** [comment]
> But interestingly, I am currently simulating a model using multiprocessing and the memory issues are indeed avoided. I can see how the memory is cleared after each subprocesses has finished its execution. So in other words, I can see how the memory increases during evaluation, but once the job is finished the process memory is killed and I don't run out of memory anymore.

Hmmm. That makes it sound like it is in Myokit. Or at least in whatever  runs inside that subprocess :-(

**Comment by DavAug** [comment]
> OK, so...
> 
> 1. There's a memory leak in Myokit and we're testing correctly but misinterpreting the result,
> 2. There's a memory leak in Myokit but our tests aren't picking it up
> 3. There's a memory leak somewhere else
> 
> I feel like we've done a reasonable job of discounting 1, so I'll try and look into 2 a bit :D

It is definitely possible that I am creating the memory problems elsewhere. But I am only using python code (mostly) numpy, which made me think it might be myokit.

**Comment by MichaelClerx** [comment]
You could try testing a full run (numpy and all) with pympler?

**Comment by DavAug** [comment]
> You could try testing a full run (numpy and all) with pympler?

Yes I did that counting the python objects and the number of Python objects is constant. In other words, pymbler doesn't pick up on whatever is happening.

**Comment by MichaelClerx** [comment]
I've tried a few other tests like enabling logging and using fixed log times (all of which use bits of python and C code not currently touched by our tests), but doesn't show much

**Comment by MichaelClerx** [comment]
![hmmmm](https://user-images.githubusercontent.com/517644/164760666-9ecc7dbb-7bc3-4265-a56a-70ac1d0a34ff.png)

Hmmmm! OK so doing lots more repeats does show an increase!

**Comment by DavAug** [comment]
> I've tried a few other tests like enabling logging and using fixed log times (all of which use bits of python and C code not currently touched by our tests), but doesn't show much

Thanks! I am not really sure what do about this. At least it really only seems t

> ![hmmmm](https://user-images.githubusercontent.com/517644/164760666-9ecc7dbb-7bc3-4265-a56a-70ac1d0a34ff.png)
> 
> Hmmmm! OK so doing lots more repeats does show an increase!

haha I have been running the same while we were chatting 😂 Ok that is a relief!!! Because now we know myokit is the problem.

**Comment by DavAug** [comment]
> ![hmmmm](https://user-images.githubusercontent.com/517644/164760666-9ecc7dbb-7bc3-4265-a56a-70ac1d0a34ff.png)
> 
> Hmmmm! OK so doing lots more repeats does show an increase!

This is with sensitivities correct? Or without?

**Comment by MichaelClerx** [comment]
Without!

**Comment by MichaelClerx** [comment]
OK. Unless you have any really good ideas I suggest we either

1. Give up and just subprocesses
2. Explore this ticket: https://github.com/MichaelClerx/myokit/issues/871

I'll do (2) at some point, just not in the very near future...

**Comment by DavAug** [comment]
> OK. Unless you have any really good ideas I suggest we either
> 
> 1. Give up and just subprocesses
> 2. Explore this ticket: [Look into checking C extensions for memory leaks with valgrind #871](https://github.com/MichaelClerx/myokit/issues/871)
> 
> I'll do (2) at some point, just not in the very near future...

Yes I agree. I am keen to look into 871, but also need to see whether I can find time for it any time soon.

**Comment by MichaelClerx** [comment]
Some updates:

1. On mac the `ru_maxrss` value may be in bytes instead of kilobytes on linux, so numbers from scripts using `resource` should be 1000x larger on mac https://stackoverflow.com/questions/59913657/strange-values-of-get-rusage-maxrss-on-macos-and-linux Not sure though, docs are a bit conflicting on this
2. Pympler does not notice if I leak floats from the simulation. It only notices when we leak things that are tracked by `gc`, or things that are referenced by an object tracked by gc https://docs.python.org/3/library/gc.html#gc.is_tracked

**Comment by MichaelClerx** [comment]
Still seeing this, after manual checking of C code involved.
However, much much less than observed in leak fixed in https://github.com/MichaelClerx/myokit/pull/883
So going back to the assumption that this is a fragmentation issue?

**Comment by MichaelClerx** [comment]
https://github.com/myokit/myokit-examples/tree/main/technical-notes/3-3-memory-leaks

@DavAug I've done some more investigation. 

- The "leak" we see above is less than what you see if you deliberately leak 1 C double per iteration (and much less than if you deliberately leak 1 Python double per iteration). 
- It's also only a few mb on 1e5 iterations. In e.g. pints multiprocessing we only do 500 iters before starting a new process, so this issue shouldn't appear in most real life use cases?
- The initial problem you described sounds much more consistent with #883 than with the test cases we have above.

In light of all that, is it OK with you if I close this issue?
(The "test it properly with valgrind" issue will remain open at #871 )

**Comment by DavAug** [comment]
Yes perfectly ok with me 🙂 thank you for putting so much work into this!
